pymc_experimental/inference/jax_find_map.py

```diff
@@ -0,0 +1,450 @@
+from pymc.model.transform.optimization import freeze_dims_and_data
+from pymc.sampling.jax import get_jaxified_graph
+from pymc.util import get_default_varnames
+from pytensor.tensor import TensorVariable
+from scipy import stats
+from scipy.optimize import OptimizeResult
+
+_log = logging.getLogger(__name__)
+
+
+def get_near_psd(A: np.ndarray) -> np.ndarray:
+    """
+    Compute the nearest positive semi-definite matrix to a given matrix.
+
+    This function takes a square matrix and returns the nearest positive
+    semi-definite matrix using eigenvalue decomposition. It ensures all
+    eigenvalues are non-negative. The "nearest" matrix is defined in terms
+    of the Frobenius norm.
+
+    Parameters
+    ----------
+    A : np.ndarray
+        Input square matrix.
+
+    Returns
+    -------
+    np.ndarray
+        The nearest positive semi-definite matrix to the input matrix.
+    """
+    C = (A + A.T) / 2
+    eigval, eigvec = np.linalg.eig(C)
+    eigval[eigval < 0] = 0
+
+    return eigvec @ np.diag(eigval) @ eigvec.T
+
+
+def _get_unravel_rv_info(optimized_point, variables, model):
+    cursor = 0
+    slices = {}
+    out_shapes = {}
+
+    for i, var in enumerate(variables):
+        raveled_shape = np.prod(optimized_point[var.name].shape).astype(int)
+        rv = model.values_to_rvs.get(var, var)
+
+        idx = slice(cursor, cursor + raveled_shape)
+        slices[rv] = idx
+        out_shapes[rv] = tuple(
+            [len(model.coords[dim]) for dim in model.named_vars_to_dims.get(rv.name, [])]
+        )
+        cursor += raveled_shape
+
+    return slices, out_shapes
+
+
+def _create_transformed_draws(H_inv, slices, out_shapes, posterior_draws, model, chains, draws):
+    X = pt.tensor("transformed_draws", shape=(chains, draws, H_inv.shape[0]))
+    out = []
+    for rv, idx in slices.items():
+        f = model.rvs_to_transforms[rv]
+        untransformed_X = f.backward(X[..., idx]) if f is not None else X[..., idx]
+
+        if rv in out_shapes:
+            new_shape = (chains, draws) + out_shapes[rv]
+            untransformed_X = untransformed_X.reshape(new_shape)
+
+        out.append(untransformed_X)
+
+    f_untransform = pytensor.function([X], out, mode="JAX")
+    return f_untransform(posterior_draws)
+
+
+def fit_laplace(
+    optimized_point: dict[str, np.ndarray],
+    model: pm.Model,
+    chains: int = 2,
+    draws: int = 500,
+    on_bad_cov: Literal["warn", "error", "ignore"] = "ignore",
+    transform_samples: bool = True,
+    zero_tol: float = 1e-8,
+    diag_jitter: float | None = 1e-8,
+    progressbar: bool = True,
+    mode: str = "JAX",
+) -> az.InferenceData:
+    """
+    Compute the Laplace approximation of the posterior distribution.
+
+    The posterior distribution will be approximated as a Gaussian
+    distribution centered at the posterior mode.
+    The covariance is the inverse of the negative Hessian matrix of
+    the log-posterior evaluated at the mode.
```

Comment:
? Less ambiguous since there is a `mode` argument
```suggestion
    the log-posterior evaluated at the optimized_point.
```